src/database/sql_database.cc

```diff
@@ -624,69 +622,64 @@ std::vector<std::shared_ptr<SQLDatabase::AddUpdateTable>> SQLDatabase::_addUpdat
+                cdsObjectSql.emplace("location_hash", quote(stringHash(dbLocation.string())));
             } else {
                 // URLs
-                cdsObjectSql["location"] = quote(loc);
-                cdsObjectSql["location_hash"] = SQL_NULL;
-            }
-        } else {
-            if (op == Operation::Update) {
-                cdsObjectSql["location"] = SQL_NULL;
-                cdsObjectSql["location_hash"] = SQL_NULL;
+                cdsObjectSql.emplace("location", quote(loc));
+                cdsObjectSql.emplace("location_hash", SQL_NULL);
             }
+        } else if (op == Operation::Update) {
+            cdsObjectSql.emplace("location", SQL_NULL);
+            cdsObjectSql.emplace("location_hash", SQL_NULL);
         }
 
         if (item->getTrackNumber() > 0) {
-            cdsObjectSql["track_number"] = quote(item->getTrackNumber());
-        } else {
-            if (op == Operation::Update)
-                cdsObjectSql["track_number"] = SQL_NULL;
+            cdsObjectSql.emplace("track_number", quote(item->getTrackNumber()));
+        } else if (op == Operation::Update) {
+            cdsObjectSql.emplace("track_number", SQL_NULL);
         }
 
-        cdsObjectSql["bookmark_pos"] = quote(item->getBookMarkPos().count());
-
         if (item->getPartNumber() > 0) {
-            cdsObjectSql["part_number"] = quote(item->getPartNumber());
-        } else {
-            if (op == Operation::Update)
-                cdsObjectSql["part_number"] = SQL_NULL;
+            cdsObjectSql.emplace("part_number", quote(item->getPartNumber()));
+        } else if (op == Operation::Update) {
+            cdsObjectSql.emplace("part_number", SQL_NULL);
         }
 
+        cdsObjectSql.emplace("bookmark_pos", quote(item->getBookMarkPos().count()));
+
         if (!item->getServiceID().empty()) {
             if (!hasReference || std::static_pointer_cast<CdsItem>(refObj)->getServiceID() != item->getServiceID())
-                cdsObjectSql["service_id"] = quote(item->getServiceID());
+                cdsObjectSql.emplace("service_id", quote(item->getServiceID()));
             else
-                cdsObjectSql["service_id"] = SQL_NULL;
-        } else {
-            if (op == Operation::Update)
-                cdsObjectSql["service_id"] = SQL_NULL;
+                cdsObjectSql.emplace("service_id", SQL_NULL);
+        } else if (op == Operation::Update) {
+            cdsObjectSql.emplace("service_id", SQL_NULL);
         }
 
-        cdsObjectSql["mime_type"] = quote(item->getMimeType().substr(0, 40));
+        cdsObjectSql.emplace("mime_type", quote(item->getMimeType().substr(0, 40)));
     }
 
+    if (obj->getParentID() == INVALID_OBJECT_ID) {
+        throw_std_runtime_error("Tried to create or update an object {} with an illegal parent id {}", obj->getLocation().c_str(), obj->getParentID());
+    }
+
+    cdsObjectSql.emplace("parent_id", quote(obj->getParentID()));
+
```

Comment:
Moving this up, may trigger the exception for an entry which could be skipped by the following check (`return {}`).

Reply:
I had this in mind, too. But since we filter for the same `parent_id`, finding such an item would mean that an illegal object was inserted before and the database is invalid. If this case should be ignored (by evaluating the invalid id after duplicated checking), I can revert the code-move. Otherwise I prefer a simple check before the more expensive SQL query.

Reply:
First of all: there should be none of this. If duplicate entries are thrown in to be added, they may have broken `parent_id`.

Errata: If `parent_id` is _invalid_ there should not be a duplicate entry for this, because it could never have made its way into the database, because of the same check.